network/rcmgr.go

```diff
@@ -0,0 +1,152 @@
+package network
+
+import (
+	"errors"
+
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+)
+
+var (
+	ErrResourceLimitExceeded = errors.New("resource limit exceeded")
+	ErrResourceScopeClosed   = errors.New("resource scope closed")
+)
+
+// ResourceManager is the interface to the network resource management subsystem
+type ResourceManager interface {
+	// Viewystem views the system wide resource scope
+	ViewSystem(func(ResourceScope) error) error
+	// ViewTransient views the transient (DMZ) resource scope
+	ViewTransient(func(ResourceScope) error) error
+	// ViewService retrieves a service-specific scope
+	ViewService(string, func(ServiceScope) error) error
+	// ViewProtocol views the resource management scope for a specific protocol.
+	ViewProtocol(protocol.ID, func(ProtocolScope) error) error
+	// ViewPeer views the resource management scope for a specific peer.
+	ViewPeer(peer.ID, func(PeerScope) error) error
+
+	// OpenConnection creates a new connection scope not yet associated with any peer; the connection
+	// is scoped at the transient scope.
+	OpenConnection(dir Direction, usefd bool) (ConnectionScope, error)
+
+	// OpenStream creates a new stream scope, initially unnegotiated.
+	// An unnegotiated stream will be initially unattached to any protocol scope
+	// and constrained by the transient scope.
+	OpenStream(p peer.ID, dir Direction) (StreamScope, error)
+
+	// Close closes the resource manager
+	Close() error
+}
+
+// ResourceScope is the interface for all scopes.
+type ResourceScope interface {
+	// ReserveMemory reserves memory/buffer space in the scope.
+	ReserveMemory(size int) error
+	// ReleaseMemory explicitly releases memory previously reserved with ReserveMemory
+	ReleaseMemory(size int)
+
+	// GetBuffer reserves memory and allocates a buffer through the buffer pool.
+	GetBuffer(size int) (Buffer, error)
+
+	// Stat retrieves current resource usage for the scope.
+	Stat() ScopeStat
+
+	// BeginTxn creates a new transactional scope rooted at this scope
+	BeginTxn() (TransactionalScope, error)
+}
+
+// Buffer is a wrapper for a buffer managed by a resource scope
+type Buffer interface {
```

Comment:
Should we make this a
```go
type Buffer []byte
```

I wish we could switch out the `append` function that Go uses!

Reply:
we can't, we can't hash or compare a slice, which makes it impossible to track  :(:(

Reply:
RIP buffers.